File: model_elliptic_2d.py
import numpy as np
import sympy as sym
import fenics as fen
import matplotlib
import matplotlib.pyplot as plt

import lib_inverse_problem
import lib_plotters

matplotlib.rc('figure', figsize=(14, 8))
matplotlib.rc('figure.subplot', hspace=.1)
matplotlib.rc('font', size=14)
matplotlib.rc('savefig', bbox='tight')
matplotlib.rc('text', usetex=True)

# Fix seed
np.random.seed(20)

# Definition of the forward model {{{1
# ====================================


class ForwardDarcy():

    def __init__(self, N, K):
        # Variables
        x, y = sym.symbols('x[0], x[1]', real=True, positive=True)
        f = sym.Function('f')(x, y)

        # Precision (inverse covariance) operator, when α = 1
        tau, alpha = 3, 2
        precision = (- sym.diff(f, x, x) - sym.diff(f, y, y) + tau**2*f)

        indices = [(m, n) for m in range(0, N) for n in range(0, N)]
        # indices = indices[:-1]
        self.indices = sorted(indices, key=sum)
        eig_f = [sym.cos(i[0]*sym.pi*x) * sym.cos(i[1]*sym.pi*y)
                 for i in self.indices]

        # Eigenvalues of the covariance operator
        eig_v = [1/(precision.subs(f, e).doit()/e).simplify()**alpha
                 for e in eig_f]

        grid = np.linspace(0, 1, K + 2)[1:-1]
        x_obs, y_obs = np.meshgrid(grid, grid)
        self.x_obs, self.y_obs = x_obs.reshape(K*K), y_obs.reshape(K*K)

        # Basis_functions
        self.functions = [f*np.sqrt(float(v)) for f, v in zip(eig_f, eig_v)]

        # --- THIS PART USED TO NOT BE PARALELLIZABLE --- #
        # Create mesh and define function space
        n_mesh = 80
        mesh = fen.UnitSquareMesh(n_mesh, n_mesh)
        self.f_space = fen.FunctionSpace(mesh, 'P', 2)

        # Define boundary condition
        # u_boundary = fen.Expression('x[0] + x[1]', degree=2)
        u_boundary = fen.Expression('0', degree=2)
        self.bound_cond = fen.DirichletBC(
           self.f_space, u_boundary,
           lambda x, on_boundary: on_boundary)

        # Define variational problem
        self.trial_f = fen.TrialFunction(self.f_space)
        self.test_f = fen.TestFunction(self.f_space)
        rhs = fen.Constant(50)
        # rhs = fen.Expression('sin(x[0])*sin(x[1])', degree=2)
        self.lin_func = rhs*self.test_f*fen.dx
        # ----------------------------------------------- #

    def __call__(self, u_test, return_sol=False, return_permeability=False):

        # Pad with zeros
        u_test = np.pad(u_test, (0, len(self.indices) - len(u_test)))

        # Assembling diffusivity
        log_coeff = sum([ui*fi for ui, fi in zip(u_test, self.functions)], 0)
        if return_permeability:
            return log_coeff

        ccode_coeff = sym.ccode(sym.exp(log_coeff))
        diff = fen.Expression(ccode_coeff, degree=2)

        # Define bilinear form in variational problem
        a_bil = diff*fen.dot(fen.grad(self.trial_f),
                             fen.grad(self.test_f))*fen.dx

        # Compute solution
        sol = fen.Function(self.f_space)
        fen.solve(a_bil == self.lin_func, sol, self.bound_cond)

        evaluations = [sol(xi, yi) for xi, yi in zip(self.x_obs, self.y_obs)]
        # print("Evaluating G...:", np.array(evaluations))
        return sol if return_sol else np.array(evaluations)


# Test code {{{1
# ==============
if __name__ == "__main__":
    G = ForwardDarcy(4, 10)
    u_truth = np.random.randn(len(G.indices))
    solution = G(u_truth, return_sol=True)
    p = fen.plot(solution)
    plt.colorbar(p)
    plt.show()

# Parameters of the inverse problem {{{1
# ======================================

# Covariance of noise and prior
γ, σ = .01, 1

# Initialize forward model
dx, Kx = 6, 10
G = ForwardDarcy(dx, Kx)

# Fourier coefficients of the true (scalar) diffusivity
u = σ*np.random.randn(len(G.indices))

# Forward model
forward = G.__call__

# Observation without noise
y = forward(u)
K = len(y)

# Forward model for approximation
dx, Kx = 3, 10
G = ForwardDarcy(dx, Kx)

# Approximation
d = len(G.indices)
u_truth = u
u = u[:d]

# Covariance of noise and prior
Σ = np.diag([σ**2]*d)
Γ = np.diag([γ**2]*K)

# Square root of covariance matrix
rtΓ = np.diag([γ]*K)

# Noisy observation
y = y + rtΓ.dot(np.random.randn(K))

# Inverse problem
ip = lib_inverse_problem.InverseProblem(forward, Γ, Σ, y, unknown=u)

# Plotters {{{1
# ==========


class MainModesPlotter(lib_plotters.MainModesPlotter):

    def __init__(self, ip, **config):
        super().__init__(ip, **config)
        ix = [G.indices[c] for c in self.coeffs]
        ax0, ax1 = self.ax[0], self.ax[1]
        ax0.set_title("Fourier coefficients {} and {}".format(ix[0], ix[1]))
        ax1.set_title("Fourier coefficients {} and {}".format(ix[0], ix[2]))
        ax0.set_xlabel("Coefficient {}".format(ix[0]))
        ax0.set_ylabel("Coefficient {}".format(ix[1]))
        ax1.set_xlabel("Coefficient {}".format(ix[0]))
        ax1.set_ylabel("Coefficient {}".format(ix[2]))
        ax0.set_xlim([-6, 6])
        ax0.set_ylim([-6, 6])
        ax1.set_xlim([-6, 6])
        ax1.set_ylim([-6, 6])


class AllCoeffsPlotter(lib_plotters.AllCoeffsPlotter):

    def plot_all_coeffs(self, iteration, data):
        self.ax.clear()
        super().plot(iteration, data)
        self.ax.set_xticks(np.arange(len(G.indices)))
        self.ax.set_xticklabels((str(i) for i in G.indices))


# Delete local variables to make them unaccessible outside
del γ, σ, dx, Kx, u, d, forward, Σ, Γ, rtΓ, y

File: model_elliptic_2d.jl

module ModelElliptic2d

using Gridap
import Random
import LinearAlgebra
la = LinearAlgebra

include("lib_inverse_problem.jl")
export utruth, least_squares, d, y, noise_cov, prior_cov

N = 3
# N = 1
indices = [(m, n) for m in 0:N for n in 0:N];
indices = sort(indices, by=x->(maximum(x), sum(x), x[1]))

function assemble_diffusivity(u)
    # Eigenvalues of the covariance operator
    τ, α = 3, 2
    eig_v = [(π^2*norm(i)^2 + τ^2)^(-α) for i in indices]
    function diffusivity(x)
        result = 0.
        for (i, ind) in enumerate(indices)
            eigf = cos(ind[1]*π*x[1]) * cos(ind[2]*π*x[2])
            result += u[i] * sqrt(eig_v[i]) * eigf
        end
        return exp(result)
    end
    return diffusivity
end

d = length(indices)
utruth = zeros(d) .+ 1.

rhs(x) = 50
domain = (0,1,0,1)
partition = (100,100)
# partition = (20,20)
model = CartesianDiscreteModel(domain,partition)

order = 1
reffe = ReferenceFE(lagrangian,Float64,order)
V0 = TestFESpace(model,reffe,conformity=:H1,dirichlet_tags="boundary")
U = TrialFESpace(V0, x -> 0)

degree = 2
Ω = Triangulation(model)
dΩ = Measure(Ω,degree)

# Find node indices of observations
function get_obs_nodes()
    M = 10
    grid = 1/M * (1:M-1) |> collect
    xobs = grid * ones(M-1)'
    yobs = ones(M-1)  * grid'
    xobs = reshape(xobs, ((M - 1)^2, 1))
    yobs = reshape(yobs, ((M - 1)^2, 1))
    nodes = Ω.node_coords

    obs_nodes = zeros(Int, (M-1)^2)
    iobs = 1

    for (inode, node) in enumerate(nodes)
        if iobs > length(obs_nodes)
            return obs_nodes
        end
        vec = [node[1], node[2]]
        obs = [xobs[iobs], yobs[iobs]]
        if norm(vec - obs) < 1e-10
            obs_nodes[iobs] = inode
            iobs += 1
        end
    end
end

# Evaluate finite element function at nodes
function eval_at_nodes(uh)
    nodes = Ω.node_coords
    elements = Ω.cell_node_ids;
    dof_values = uh.cell_dof_values
    node_values = zeros(length(nodes))
    for (ielem, dof_values) in enumerate(dof_values)
        for (idof, dofvalue) in enumerate(dof_values)
            inode = elements[ielem][idof]
            node_values[inode] = dofvalue
        end
    end
    return node_values
end

obs_nodes = get_obs_nodes()
function forward(u)
    if length(u) != length(indices)
        println("Wrong input!")
    end
    diff = assemble_diffusivity(u)
    a(u,v) = ∫( diff * ∇(v) ⋅ ∇(u) )*dΩ
    b(v) = ∫( v*rhs )*dΩ
    op = AffineFEOperator(a,b,U,V0)
    uh = solve(op)
    uh_nodes = eval_at_nodes(uh)
    uh_obs = uh_nodes[obs_nodes]
end

# Covariance of noise and prior
γ, σ = .01, 1

# Truth
Random.seed!(3)
utruth = σ*randn(d)
y = forward(utruth)

# Square root of covariance matrix
K = length(y)
rtΓ = la.diagm(γ .+ zeros(K))

# Noisy observation
y = y + rtΓ*randn(K)

# Covariance matrices
noise_cov = la.diagm(γ^2 .+ zeros(K))
prior_cov = la.diagm(σ^2 .+ zeros(d))

ip = Ip.InverseProblem(forward, y, noise_cov, prior_cov)
end
